### PR TITLE
Sigv4a

### PR DESCRIPTION
**Pull request by bretambrose**
* Sigv4a implementation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: awscrt/auth.py [hunk 1]
```diff
@@ -401,9 +404,13 @@ class AwsSigningConfig(NativeResource):
             computed from the signable.
 
         credentials_provider (AwsCredentialsProvider): Credentials provider
-            to fetch signing credentials with.
+            to fetch signing credentials with. If the signing algorithm is
+            V4_ASYMMETRIC, ecc-based credentials will be derived from the
```
Comment (graebm) [inline awscrt/auth.py:408]: Turn this into a link like so (note the backticks):
```
`:attr:AwsSigningAlgorithm.V4_ASYMMETRIC`
```

Comment (graebm) [reply awscrt/auth.py:408]: also: capitalize ECC

File: awscrt/auth.py [hunk 2]
```diff
@@ -550,12 +557,20 @@ def signature_type(self):
 
     @property
     def credentials_provider(self):
-        """AwsCredentialsProvider: Credentials provider to fetch signing credentials with"""
+        """
+        AwsCredentialsProvider: Credentials provider to fetch signing credentials with.
+        If the signing algorithm is V4_ASYMMETRIC, ecc-based credentials will be derived
```
Comment (graebm) [inline awscrt/auth.py:562]: likewise, turn V4_ASYMMETRIC into link and capitalize ECC

File: awscrt/auth.py [hunk 3]
```diff
@@ -401,9 +404,13 @@ class AwsSigningConfig(NativeResource):
...
 
         credentials_provider (AwsCredentialsProvider): Credentials provider
-            to fetch signing credentials with.
+            to fetch signing credentials with. If the signing algorithm is
+            V4_ASYMMETRIC, ecc-based credentials will be derived from the
+            fetched credentials.
 
-        region (str): The region to sign against.
+        region (str): If signing algorithm is V4, the region to sign against.
+            If signing algorithm is V4_ASYMMETRIC, the value of the
```
Comment (graebm) [inline awscrt/auth.py:412]: links for V4 and V4_ASYMMETRIC

File: awscrt/auth.py [hunk 4]
```diff
@@ -401,9 +404,13 @@ class AwsSigningConfig(NativeResource):
...
+            X-amzn-region-set header (added in signing).
```
Comment (graebm) [inline awscrt/auth.py:413]: trivial: quotes around "X-amzn-region-set" we do that when referencing other headers/query-params

File: awscrt/auth.py [hunk 5]
```diff
@@ -550,12 +557,20 @@ def signature_type(self):
...
+        """
         return _awscrt.signing_config_get_credentials_provider(self._binding)
 
     @property
     def region(self):
-        """str: The region to sign against"""
+        """
+        str: If signing algorithm is V4, the region to sign against.
+        If signing algorithm is V4_ASYMMETRIC, the value of the
+        X-amzn-region-set header (added in signing).
```
Comment (graebm) [inline awscrt/auth.py:572]: same as above, links and quotes around header name

File: source/auth_signer.c [hunk 6]
```diff
@@ -105,6 +105,7 @@ PyObject *aws_py_sign_request_aws(PyObject *self, PyObject *args) {
 
     async_data->signable = aws_signable_new_http_request(aws_py_get_allocator(), http_request);
     if (!async_data->signable) {
+        PyErr_SetAwsLastError();
```
Comment (graebm) [inline source/auth_signer.c:108]: 😱
great catch

**Review by graebm — COMMENTED** [review]
re-generate the docs too, if you're able

**Review by graebm — APPROVED** [review]
fix & ship

**Comment by TingDaoK** [comment]
have no `aws_verify_sigv4a_signing` binding for python... Do we need it?

File: source/auth_signer.c (hunk 6 above)
Comment (bretambrose) [reply source/auth_signer.c:108]: git blame says you added this!

**Comment by bretambrose** [comment]
> have no `aws_verify_sigv4a_signing` binding for python... Do we need it?



> have no `aws_verify_sigv4a_signing` binding for python... Do we need it?

we'd need it if we checked the actual signature value in the test, but we don't

**Pull request merged**